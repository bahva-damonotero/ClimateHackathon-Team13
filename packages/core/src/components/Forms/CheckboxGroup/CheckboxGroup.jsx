import React from 'react';
import PropTypes from 'prop-types';

import { flattenChildren } from '@boozallen-uip/uswds-react-utils';

import { Fieldset, FormError } from '..';

export default function CheckboxGroup({ children, error, id, tile }) {
  let label;
  const options = [];

  flattenChildren(React.Children.toArray(children)).forEach((child) => {
    const componentType = child.type.displayName;

    switch (componentType) {
      case 'Legend':
        label = child;
        break;

      case 'Checkbox':
        options.push(child);
        break;

      default:
        throw new Error(
          `Invalid child component '${componentType}' used in CheckboxGroup.`
        );
    }
  });

  if (!label) {
    throw new Error('A Legend is required inside of a CheckboxGroup.');
  }

  if (!options.length) {
    throw new Error(
      'At least one Checkbox must be present inside of a CheckboxGroup.'
    );
  }

  return (
    <Fieldset>
      {React.cloneElement(label, { error })}

      <FormError error={error} />

      {options.map((child, index) =>
        React.cloneElement(child, {
          groupIndex: index,
          id,
          tile
        })
      )}
    </Fieldset>
  );
}

CheckboxGroup.propTypes = {
  /** `<Legend>` + `<Checkbox>` node(s) */
  children: PropTypes.node.isRequired,
  /** Error message to display */
  error: PropTypes.string,
  /** ID for the group of checkboxes */
  id: PropTypes.string.isRequired,
  /** If `true`, the tile checkbox variant is used for the entire group */
  tile: PropTypes.bool
};

CheckboxGroup.defaultProps = {
  tile: false
};
