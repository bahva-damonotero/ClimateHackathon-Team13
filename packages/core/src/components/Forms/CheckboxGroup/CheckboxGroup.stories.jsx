/* eslint-disable react/prop-types */
import React from 'react';
import { useState } from '@storybook/addons';

import CheckboxGroup from '.';
import { Checkbox, CheckboxLabelDescription, Form, Legend } from '..';

export default {
  title: 'Forms/CheckboxGroup',
  component: CheckboxGroup,
  subcomponents: {
    Legend,
    Checkbox,
    CheckboxLabelDescription
  },
  argTypes: {
    label: {
      name: '<Legend> children',
      description: 'Label for the field',
      type: { required: true },
      table: {
        type: { summary: 'node' }
      }
    },
    screenReaderOnly: {
      name: '<Legend> screenReaderOnly',
      description:
        'If `true`, the label will only be visible to screen readers.',
      control: {
        type: 'boolean'
      },
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'false' }
      }
    },
    children: {
      control: { type: null }
    }
  }
};

const defaultArgs = {
  id: 'checkbox-group-example',
  label: 'Label'
};

const DefaultTemplate = ({ label, screenReaderOnly, ...args }) => {
  const [choiceOne, setChoiceOne] = useState(false);
  const [choiceTwo, setChoiceTwo] = useState(false);

  return (
    <Form>
      <CheckboxGroup {...args}>
        <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

        <Checkbox
          checked={choiceOne}
          onChange={(e) => setChoiceOne(e.target.checked)}
        >
          Choice One
        </Checkbox>

        <Checkbox
          checked={choiceTwo}
          onChange={(e) => setChoiceTwo(e.target.checked)}
        >
          Choice Two
        </Checkbox>

        <Checkbox disabled>Choice Three</Checkbox>
      </CheckboxGroup>
    </Form>
  );
};

export const def = DefaultTemplate.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

const SelectionsTemplate = ({ label, screenReaderOnly, ...args }) => (
  <Form>
    <CheckboxGroup {...args}>
      <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

      <Checkbox checked readOnly>
        Choice One
      </Checkbox>

      <Checkbox readOnly>Choice Two</Checkbox>

      <Checkbox checked disabled>
        Choice Three
      </Checkbox>
    </CheckboxGroup>
  </Form>
);

export const selections = SelectionsTemplate.bind({});
selections.args = {
  ...defaultArgs
};

export const error = SelectionsTemplate.bind({});
error.args = {
  ...defaultArgs,
  error: 'Please check all 3 boxes'
};

export const tile = DefaultTemplate.bind({});
tile.args = {
  ...defaultArgs,
  tile: true
};

const TileWithOptionalTextTemplate = ({ label, screenReaderOnly, ...args }) => {
  const [choiceOne, setChoiceOne] = useState(false);
  const [choiceTwo, setChoiceTwo] = useState(false);

  return (
    <Form>
      <CheckboxGroup {...args}>
        <Legend screenReaderOnly={screenReaderOnly}>{label}</Legend>

        <Checkbox
          checked={choiceOne}
          onChange={(e) => setChoiceOne(e.target.checked)}
        >
          Choice One
        </Checkbox>

        <Checkbox
          checked={choiceTwo}
          onChange={(e) => setChoiceTwo(e.target.checked)}
        >
          Choice Two
          <CheckboxLabelDescription>
            This is optional text that can be used to describe the label in more
            detail.
          </CheckboxLabelDescription>
        </Checkbox>

        <Checkbox disabled>Choice Three</Checkbox>
      </CheckboxGroup>
    </Form>
  );
};

export const tileWithOptionalText = TileWithOptionalTextTemplate.bind({});
tileWithOptionalText.args = {
  ...defaultArgs,
  tile: true
};
