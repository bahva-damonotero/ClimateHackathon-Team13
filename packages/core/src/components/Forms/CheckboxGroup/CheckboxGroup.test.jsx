import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import CheckboxGroup from '.';
import { Checkbox, Legend } from '..';

describe('CheckboxGroup', () => {
  const id = 'checkbox-group-example';
  const label = 'Checkbox Group Label';

  const jsx = (
    <CheckboxGroup id={id}>
      <Legend>{label}</Legend>

      <Checkbox readOnly>Choice One</Checkbox>

      <Checkbox readOnly>Choice Two</Checkbox>

      <Checkbox disabled>Choice Three</Checkbox>
    </CheckboxGroup>
  );

  it('should render correctly', () => {
    const { container } = render(jsx);

    expect(container.querySelector('.usa-fieldset .usa-legend').innerHTML).toBe(
      label
    );
  });

  it('should throw an error if a Legend is not provided', () => {
    expect(() => {
      render(
        <CheckboxGroup id={id}>
          <Checkbox readOnly>Choice One</Checkbox>

          <Checkbox readOnly>Choice Two</Checkbox>

          <Checkbox disabled>Choice Three</Checkbox>
        </CheckboxGroup>
      );
    }).toThrow('A Legend is required inside of a CheckboxGroup.');
  });

  it('should throw an error if no Checkbox components are provided', () => {
    expect(() => {
      render(
        <CheckboxGroup id={id}>
          <Legend>{label}</Legend>
        </CheckboxGroup>
      );
    }).toThrow(
      'At least one Checkbox must be present inside of a CheckboxGroup.'
    );
  });

  it('should throw an error when an invalid component is used in the CheckboxGroup', () => {
    expect(() => {
      render(
        <CheckboxGroup id={id}>
          <Legend>{label}</Legend>

          <Checkbox readOnly>Choice One</Checkbox>

          <Checkbox readOnly>Choice Two</Checkbox>

          <Checkbox disabled>Choice Three</Checkbox>

          <span>Some content that doesn't belong here.</span>
        </CheckboxGroup>
      );
    }).toThrow("Invalid child component 'undefined' used in CheckboxGroup.");
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
