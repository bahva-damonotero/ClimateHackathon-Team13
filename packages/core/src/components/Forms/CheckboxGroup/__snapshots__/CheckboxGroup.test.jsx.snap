// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup should match snapshot 1`] = `
<fieldset
  className="usa-fieldset"
>
  <legend
    className="usa-legend"
  >
    Checkbox Group Label
  </legend>
  <div
    className="usa-checkbox"
  >
    <input
      checked={false}
      className="usa-checkbox__input"
      disabled={false}
      id="checkbox-group-example--0"
      name="checkbox-group-example"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="usa-checkbox__label"
      htmlFor="checkbox-group-example--0"
    >
      Choice One
    </label>
  </div>
  <div
    className="usa-checkbox"
  >
    <input
      checked={false}
      className="usa-checkbox__input"
      disabled={false}
      id="checkbox-group-example--1"
      name="checkbox-group-example"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="usa-checkbox__label"
      htmlFor="checkbox-group-example--1"
    >
      Choice Two
    </label>
  </div>
  <div
    className="usa-checkbox"
  >
    <input
      checked={false}
      className="usa-checkbox__input"
      disabled={true}
      id="checkbox-group-example--2"
      name="checkbox-group-example"
      readOnly={false}
      type="checkbox"
    />
    <label
      className="usa-checkbox__label"
      htmlFor="checkbox-group-example--2"
    >
      Choice Three
    </label>
  </div>
</fieldset>
`;
